New RFC for CAPI Kubernetes/OS version management (#54)
giantswarm/rfc: capi-matrix/README.md

# RFC 0015 - CAPI matrix per provider

## Introduction

We have a cluster-<provider> chart for every provider that we support. It has a set of templates that allow our customers to create a cluster app with all necessary resources required by a Cluster API implementation. 

In the values of these charts we define the Kubernetes version of the cluster and the image that machine will use to run. These images contain the Kubernetes components with a specific version. 

> One of the goal in this RFC is decouple Kubernetes version from image names so customer does not need to know which image correspond to which version. At the same time we might want to provide new OS version when new CVE appears without necessarily change the Kubernetes version. Finally having a list of kubernetes version certified as ready to be used benefit to disallow users to create cluster in flawy versions.

Reviewer: > flawy

Typo?

## User stories

- As a customer, I want to not worry about the kubernetes version and rely on Giant Swarm default version provided.

- As Giant Swarm staff, I don't want to expend time upgrading cluster manually.

- As Giant Swarm staff, I would like to forbid a specific Kubernetes/OS version to be used in case of security constraints.

> - As customer, I want to have the option to have a cluster pinned to a version.

Reviewer: Can we include reasons here? 

And what boundaries? 
Will it be pin to a specific kubernetes/OS combo? Pinned to a patch? Pinned to a major?
Will we let them pin for 6 months? 1 year?

Author: sure

## Solution

> - Lets have a cluster-<provider> chart release for every Kubernetes version is released. That way our automation can roll upgrades automatically when new version is published

Reviewer: I think this is the wrong approach.

I _still_ think we need an abstraction on top of cluster apps that the user deals with. That layer should include this kind of pinning.

Author: We proposed a meta app in the past and I think we discussed it in Kass sync. There were some drawbacks and it was discarded for now. What is the problem of using a kuberneteVersion default in this layer? BTW CAPG and CAPA is doing too

Author: but this is not limiting or changing anything about your approach. It just makes easier the customer life, defaulting the image name based on the kubernetes version (at the same time allow us to control which versions are available to the customer)

Reviewer 2: What Marcus is saying is that there should be just one version available. So neither the customer should be able to pin any version nor we provide any choice. I'd like to get there too. 

But how would we solve freeze periods then? Do we have confidence that we won't break anything and everything is just kept up to date whatsoever? Or do we still need to be able to stay at a certain version? And will we be able to test the upgrade that happens after the freeze period? This might be a bigger jump.

Author: IMO forcing a version per cluster_PROVIDER release limits us not the customer. Let me explain it. In our whole history we have been managing clusters and apps, and in many times across all customers we have some requirements, blockers,... that block us to upgrade to certain versions of an app or a k8s (even OS when cgroups).

This is going to happen no matter how good the cluster upgrade mechanics are. You also mention other conditions like frozen periods or software dependencies (openet ehem ehem) that force us to offer that option.

IMO the problem we had in the past is the manual toil of schedule, discuss, manual preparation, ... which required us to do an upgrade. This is what we need to solve and for that reason [the automate cluster upgrade rfc](https://github.com/giantswarm/rfc/tree/main/automatic-cluster-upgrades) was created. 

I don't say that we need to put the list of supported versions in cluster-PROVIDER charts, we could find a better place, but we will need that list somewhere. Leaving the chart without the default k8s/OS version and defining it in another place for all providers is another option. But we need to clarify the vision of how clusters/app should be updated.

- Lets include a matrix between Kubernetes version and images names in the chart.

```yaml

> kubernetesVersionImages:

Reviewer: Is this useful to the customer? How will that know what to pick? Why would we encourage the use of old (unsupported) Kubernetes versions?

Author: No this is not useful for the end customer (maybe there is a corner case) but maybe for other implementators like I describe at bottom of the doc. I have doubted to embedded in the _helpers but setting it values allow us to change it in real live if a critical issue appears

Reviewer 3: I still don't really get why there needs to be a list of versions, if every new OS/K8s version is anyway a new chart release/version. Wouldn't each chart release then only need a single image?

I thought we pin an OS/K8s combination per chart release. This way the upgrade is moving to the next chart version, which is something that is easy to automate also later based on upstream automation instead of having to rely on value fields.

Only exception I'd see and that I'd like to support is having different images for CP and worker nodes, cause sometimes we might just want to update the CP nodes (CVE in K8s CP) without having to roll all workers.

Reviewer 2: This would make the lifecycle of kubernetes independent from other features within the cluster chart. Eg independent of bastions, private clusters etc. But some other parts in the chart are still bound to kubernetes too. So if we want to have a team responsible for kubernetes itself and other teams responsible for infrastructure integration in a specific infrastructure like aws or openstack, how can we split responsibility within the same chart? Does the kubernetes team then still need to do releases of each provider chart? And is that a desired design? Are we even able to separate this completely? We need to make sure we separate the configuration for k8s itself and the provider one. The k8s team still needs to run tests on all providers and they need to have changes in their own chart dependent on which provider. But without the need to update all charts that are maintained by the infrastructure teams. They need to be able to roll out their own changes to all providers.

Another idea (not sure if discussed already) would be to allow customers to select something like beta, stable, long-term moving targets. This way teams can release their changes and the release gets tested further along and promoted through the stages.

Reviewer 4: I agree it should be hidden from the customer. Did we ever had a critical issue requiring us to change the OS so quickly we wouldn't be able to change a few lines in the chart and release it?

Author: No I don't think so. It can be hidden of course, I have a discussion with someone else and making it part of values seemed logical for other possible users (not GS). 

And Timo I expected that if there is a K8s team releasing new version there is an automated process that creates the release for each provider (anyway so far is something we do manually, so this is not getting worse). Having multiple release channels could be a solution, but I would not make it hard to select a specific version of k8s/OS since it will could impact us in the future having to make workaround to pin versions.

> # RFC 0015 - CAPI matrix per provider

Reviewer 5: We don't do numbering of RFCs, at least based on the main README.

```suggestion
# CAPI matrix per provider
```

## Introduction

> We have a cluster-<provider> chart for every provider that we support (it is our UI to the customer). It has a set of templates that allow our customers to create a cluster app with all necessary resources required by a Cluster API implementation. 

Reviewer 5: "it is our UI to the customer" sounds like this is THE user interface, as if there are no others.

I guess you'd like to express something like: The cluster app is our technical interface to configure clusters.

In terms of terminology, I would replace "chart" with "app".

Author: yeah good point. Adding it

> We have a cluster-<provider> chart for every provider that we support (it is our UI to the customer). It has a set of templates that allow our customers to create a cluster app with all necessary resources required by a Cluster API implementation. 

Reviewer 5: The term `<provider>` disappears when rendered as HTML. Changing the formatting.

```suggestion
We have a cluster-PROVIDER chart for every provider that we support (it is our UI to the customer). It has a set of templates that allow our customers to create a cluster app with all necessary resources required by a Cluster API implementation. 
```

In the values of these charts, among all possible parameters we define the Kubernetes version of the cluster and the image that machines will use to run. These images contain the Kubernetes components with a specific version. 

> This RFC born out of necessity to automate the cluster upgrades. For that reason team rocket has discussed and found a solution adding a Kubernetes and OS version in each chart release. Today AWS, GCP anf Open Stack providers pinned the Kubernetes/OS version in the chart. Here we propose an extension to give more flexibility and control to Giant Swarm and the customers.

Reviewer 5: ```suggestion
This RFC born out of necessity to automate the cluster upgrades. For that reason, team Rocket has discussed and found a solution adding a Kubernetes and OS version in each chart release. Today AWS, GCP anf Open Stack providers pinned the Kubernetes/OS version in the chart. Here we propose an extension to give more flexibility and control to Giant Swarm and the customers.
```

## User stories

> - As a customer, I don`t want to worry about the kubernetes version or OS version used by my cluster machines. Just rely on the Giant Swarm default and tested version.

Reviewer 5: ```suggestion
- As a customer, I don't want to worry about the kubernetes version or OS version used by my cluster machines. Just rely on the Giant Swarm default and tested version.
```

- As a customer, I would like to choose the kubernetes version of my cluster without think which OS AMI correspond to that version.

- As Giant Swarm staff, I want to have the capacity to push new Kubernetes/OS versions and trigger the upgrade automation.

- As Giant Swarm staff, I would like to forbid a specific Kubernetes/OS version to be used in case of security constraints.

- As Giant Swarm staff, I would like to release a new OS version with the same Kubernetes version and roll all customer clusters when a security breach is detected.

## Solution

> The solution proposed is to create a cluster-<provider> chart release for every Kubernetes/OS version is released. That way will help our automation to roll upgrades automatically when new version is published.

Reviewer 5: ```suggestion
The solution proposed is to create a cluster-PROVIDER chart release for every Kubernetes/OS version is released. That way will help our automation to roll upgrades automatically when new version is published.
```

But we would like to have a list of images attached to every Kubernetes version like:

```yaml
kubernetesVersionImages:
  v1.22.9: "ubuntu-2004-kube-v1.22.9"
  v1.22.10: "ubuntu-2004-kube-v1.22.10"
  v1.23.3: "ubuntu-2004-kube-v1.23.3"
  v1.23.4: "ubuntu-2004-kube-v1.23.4"
  v1.23.5: "ubuntu-2004-kube-v1.23.5"
```

So the customers don't have to think about image names and we can remove Kubernetes/OS versions that are flawed. For example we can change OS version keeping the Kubernetes version the same:

```yaml
kubernetesVersionImages:
  v1.22.9: "ubuntu-2005-kube-v1.22.9"
  ...
```

> Every time there is a new Kubernetes version or OS version we will (automatically) create a PR to cluster-<provider> charts including the version and potentially changing the default version (if our tests run successfully).

Reviewer 5: ```suggestion
Every time there is a new Kubernetes version or OS version we will (automatically) create a PR to cluster-PROVIDER charts including the version and potentially changing the default version (if our tests run successfully).
```

At the same time if we found a Kubernetes/OS version with a security issue, we can remove the Kubernetes/OS version and enforce customers to upgrade clusters.

We have start with [Cluster API OpenStack](https://github.com/giantswarm/cluster-openstack/pull/125).

> __Bonus__: If cluster-<provider> chart is used for companies out of Giant Swarm they can use their own matrix.

Reviewer 5: ```suggestion
__Bonus__: If cluster-PROVIDER chart is used for companies out of Giant Swarm they can use their own matrix.
```

We have a cluster-PROVIDER app for every provider that we support (it is our interface to define the cluster structure and configuration). It has a set of templates that allow our customers to create a cluster app with all necessary resources required by a Cluster API implementation. 

In the values of these charts, among all possible parameters we define the Kubernetes version of the cluster and the machine image (AMI) that will be used to run the VMs. These images contain the Kubernetes components with a specific version. There are usually three AMIs defined in every provider: bastion, control-plane and node pools.

This RFC born out of necessity to automate the cluster upgrades. For that reason, team Rocket has discussed and found a solution adding a Kubernetes and OS version in each chart release. Today AWS, GCP anf Open Stack providers pinned the Kubernetes/OS version in the chart. Here we propose an extension to give more flexibility and control to Giant Swarm and the customers.

## User stories

- As a customer, I don't want to worry about the kubernetes version or OS version used by my cluster machines. Just rely on the Giant Swarm default and tested version.

> - As a customer, I would like to choose the kubernetes version of my cluster without think which OS AMI correspond to that version.

Reviewer 3: this story directly opposes the first story above

Author: yeah it is on purpose. Unfortunately, not every customer has the same mindset. Some of them want not to worry at all and some want to have finer control. I can add as `advance` customer WDYT?

Reviewer 3: My issue with this is that we have not decided to give customers such freedom for now. The current product officially supports 2 Kubernetes minor versions at each point in time. Patch versions are not something we want to give freedom on, I would also very much question the use case here, a patch version in K8s should only fix problems and CVEs, there should be nothing speaking against it.

Furthermore, even within the 2 versions our product offering is currently:
- new features and minor changes only get into the latest
- the older version only gets patch updates

On the OS side we do not even give that choice and I would also like to keep that product decision unless there's a concrete use cases that makes sense here.

Exceptions to the above I know of, but I'd like to keep those to exceptions and not make them a feature.

TLDR I would reduce it to `As a customer, I would like to choose between the two supported minor Kubernetes versions without having to think about choosing an OS or machine image.`

> - As a customer, I don't want to worry about the kubernetes version or OS version used by my cluster machines. Just rely on the Giant Swarm default and tested version.
> 
> - As a customer, I would like to choose the kubernetes version of my cluster without think which OS AMI correspond to that version.

Author: ```suggestion
- As a customer, I would like to choose between the two supported minor Kubernetes versions without having to think about choosing an OS or machine image.
```

- As Giant Swarm staff, I want to have the capacity to push new Kubernetes/OS versions and trigger the upgrade automation.

- As Giant Swarm staff, I would like to forbid a specific Kubernetes/OS version to be used in case of security constraints.

- As Giant Swarm staff, I would like to release a new OS version with the same Kubernetes version and roll all customer clusters when a security breach is detected.

## Solution

The solution proposed is to create a cluster-PROVIDER chart release for every Kubernetes/OS version is released. That way will help our automation to roll upgrades automatically when new version is published.

But we would like to have a list of images attached to every Kubernetes version like:

```yaml

> kubernetesVersionImages:

Reviewer 6: dumb question: are these names provider specific? or would we have the same names across providers?

Author: IIRC the names are produced by our image builder CI for each OS/k8s version published. Am I right @kopiczko ?

Reviewer: Correct, the names are chosen by us.

## How the conversion image table is updated?

- We have not defined this process. But one possible scenario would be:

1) Kubernetes/Kinvolk release a new OS/k8s version

2) Our image builder job is triggered and creates the AMIs for the different providers. 

3) Other job can be triggered after and create pull requests to the cluster-PROVIDER repos adding the new k8s/OS version.

4) Our future CI pipeline can test this new version for each provider and check it runs correctly (conformance tests or even custom tests). 

5) In case CI succeed the default version and conversion image table is changed to the new one with the approval of the owning team.

From this moment Cluster Upgrade automation can start scheduling the workload cluster upgrades based on customer constraints. More info about this process discussed [here](https://github.com/giantswarm/rfc/pull/52).

## Open questions

- Every time we release a new provider app with a newer Kubernetes version we can make it default, but at the same time we might want to wait and run not edge version but a previous one. So we will need a bit of process deciding when we change the default version.

## Out of scope

> - How containerd version is managed.

Reviewer 6: if i understand correctly, we would manage containerd versions with this rfc, as they're tied to os versions

Author: then yes, but I meant here don't want to bring containerd version to the mix. I assume the containerd version will be compatible with the k8s components running on the OS

Author: But our idea is to moved Flatcar in all providers, right?
